File: backend/src/schemas/subtitle.py
"""Схемы для субтитров видео."""

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class SubtitleFormat(str, Enum):
    """Формат субтитров."""

    SRT = "srt"
    VTT = "vtt"
    ASS = "ass"


class SubtitleStatus(str, Enum):
    """Статус генерации субтитров."""

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class SubtitleCue(BaseModel):
    """Отдельная реплика субтитров."""

    id: int
    start_time: float
    end_time: float
    text: str


class GenerateSubtitlesRequest(BaseModel):
    """Запрос на генерацию субтитров."""

    video_id: uuid.UUID
    language: str = "ru"
    auto_detect: bool = True


class SubtitleResponse(BaseModel):
    """Ответ с информацией о субтитрах."""

    id: uuid.UUID
    video_id: uuid.UUID
    language: str
    status: SubtitleStatus
    format: SubtitleFormat
    cues: list[SubtitleCue] | None
    created_at: datetime

    model_config = {"from_attributes": True}


class SubtitleListResponse(BaseModel):
    """Список субтитров для видео."""

    subtitles: list[SubtitleResponse]
    video_id: uuid.UUID


class SubtitleUploadRequest(BaseModel):
    """Запрос на загрузку субтитров."""

    video_id: uuid.UUID
    language: str
    content: str
    format: SubtitleFormat = SubtitleFormat.VTT


class SubtitleSettingsRequest(BaseModel):
    """Настройки отображения субтитров."""

    enabled: bool = True
    font_size: int = Field(16, ge=10, le=32)
    background_opacity: float = Field(0.7, ge=0, le=1)
    position: str = "bottom"
    text_color: str = "#FFFFFF"
    background_color: str = "#000000"


class SubtitleSettingsResponse(BaseModel):
    """Текущие настройки субтитров."""

    enabled: bool
    font_size: int
    background_opacity: float
    position: str
    text_color: str
    background_color: str

File: frontend/src/types/subtitle.ts
/**
 * Типы для субтитров видео
 */

export type SubtitleFormat = 'srt' | 'vtt' | 'ass';
export type SubtitleStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface SubtitleCue {
    id: number;
    start_time: number;
    end_time: number;
    text: string;
}

export interface SubtitleResponse {
    id: string;
    video_id: string;
    language: string;
    status: SubtitleStatus;
    format: SubtitleFormat;
    cues: SubtitleCue[] | null;
    created_at: string;
}

export interface SubtitleSettings {
    enabled: boolean;
    font_size: number;
    background_opacity: number;
    position: 'top' | 'bottom';
    text_color: string;
    background_color: string;
}

export const DEFAULT_SUBTITLE_SETTINGS: SubtitleSettings = {
    enabled: true,
    font_size: 16,
    background_opacity: 0.7,
    position: 'bottom',
    text_color: '#FFFFFF',
    background_color: '#000000',
};

export const SUBTITLE_LANGUAGES = [
    { code: 'ru', name: 'Русский' },
    { code: 'en', name: 'English' },
    { code: 'de', name: 'Deutsch' },
    { code: 'fr', name: 'Français' },
    { code: 'es', name: 'Español' },
];
